apiserver/facades/agent/provisioner/provisioner.go

```diff
@@ -1207,13 +1207,13 @@ func (h *containerProfileHandler) ProcessOneContainer(
 		}
 
 		appName := app.Name()
-		uuid, err := h.applicationService.GetCharmIDByApplicationName(ctx, appName)
+		locator, err := h.applicationService.GetCharmLocatorByApplicationName(ctx, appName)
 		if err != nil {
 			h.SetError(idx, err)
 			return errors.Trace(err)
 		}
 
-		profile, revision, err := h.applicationService.GetCharmLXDProfile(ctx, uuid)
+		profile, revision, err := h.applicationService.GetCharmLXDProfile(ctx, locator)
```

Comment:
I don't like this.

This may have already been discussed, but imo I think it makes more sense to return the id as return value where we can. We've already interacted with state, we may as well return the densest version of a specifier

Reply:
I understand your point, I don't have a strong view on this, but:

> We've already interacted with state

here for example, we are on the apiserver, and we should have no knowledge of what happened under the service call, so if we are identifying charms with a locator (i.e. the new natural key) then we have to do it always. Otherwise the UUID is leaking through.

Reply:
What ID?

Reply:
OK, yes I agree with @jack-w-shaw.

The call should now be `profile, revision, err := h.applicationService.GetCharmLXDProfileByApplicationName(ctx, name)`

Reply:
In fact, to add to @jack-w-shaw point, we should remove or **limit** the use of `GetCharmLocatorByApplicationName`.

Reply:
I think it's fine the the UUID leaks through to users like the apiserver. The charm locator is a human reading identifier, constructable out of base parts.

The UUID is another identifier, which is opaque.

This workers with watcher for instance. The watchers will emit uuids, which the workers then look up the charm for (unless we want to re-invest charm url and serialise charm locators to a string. I will resign is we do that).

So we will likely need the methods working by id anyway

Reply:
I want to squash methods into one where possible. So if that means becoming a bit German, then so be it.